### ISSUE TITLE
Support Early Stop Mechanism

### ISSUE THREAD

**Issue by louisccc** [Sujit-O/pykg2vec#103]
Allow the trainer to stop the training if the metrics can not be improved further.

**Comment by baxtree** [comment]
Please have a look at the solution in https://github.com/Sujit-O/pykg2vec/pull/111. Thanks!

**Comment by louisccc** [comment]
related to #111, "patience" added to the trainer so it will stop the training process if the loss has not been decreasing for consecutive times.

**Comment by louisccc** [comment]
I add the idea of the interval which I learned from this paper. 

http://proceedings.mlr.press/v48/trouillon16.pdf

**Comment by baxtree** [comment]
Interesting interval-based checking. Does that mean if the loss keeps increasing, the later saved model will have higher loss? Or the model will be only saved with the best weights?

**Comment by louisccc** [comment]
it doesn't store the best weights for now. But the idea in the paper is that they check the MRR per 50 epochs and if it stops becoming better then the training process stops.

**Comment by baxtree** [comment]
Oh cool. Probably the continuous increasing in MRR should not happen on a well-designed model and general enough training sets. It would be nice to have a mechanism for trouble-shooting that.

**Comment by louisccc** [comment]
the design of the evaluator has changed to single-process, so it is easier to pass the metric back to trainier.py and judge for early stop mechanism.

**Comment by baxtree** [comment]
Hi, @louisccc. Got a new PR https://github.com/Sujit-O/pykg2vec/pull/121.

**Comment by louisccc** [comment]
new PR should close this issue. #126
